mlflow/models/evaluation/default_evaluator.py

```diff
@@ -0,0 +1,427 @@
+                pass
+
+        truncated_feature_names = [truncate_str_from_middle(f, 20) for f in feature_names]
+        for i, truncated_name in enumerate(truncated_feature_names):
+            if truncated_name != feature_names[i]:
+                # For truncated name, attach "(f_{feature_index})" at the end
+                truncated_feature_names[i] = f'{truncated_name}(f_{i})'
+
+        truncated_feature_name_map = {f: f2 for f, f2 in zip(feature_names, truncated_feature_names)}
+        if isinstance(X, pd.DataFrame):
+            X = X.rename(columns=truncated_feature_name_map, copy=False)
+
+        sampled_X = shap.sample(X, sample_rows)
+        if algorithm:
+            if algorithm == 'sampling':
+                explainer = shap.explainers.Sampling(
+                    predict_fn, X, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(X, sample_rows)
+            else:
+                explainer = shap.Explainer(
+                    predict_fn, sampled_X, feature_names=truncated_feature_names, algorithm=algorithm
+                )
+                shap_values = explainer(sampled_X)
+        else:
+            maskers = shap.maskers.Independent(sampled_X)
+            if raw_model and not is_multinomial_classifier and \
+                    shap.explainers.Linear.supports_model_with_masker(raw_model, maskers):
+                explainer = shap.explainers.Linear(
+                    raw_model, maskers, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(sampled_X)
+            elif raw_model and not is_multinomial_classifier and \
+                    shap.explainers.Tree.supports_model_with_masker(raw_model, maskers):
+                explainer = shap.explainers.Tree(
+                    raw_model, maskers, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(sampled_X)
+            elif raw_model and shap.explainers.Additive.supports_model_with_masker(
+                    raw_model, maskers
+            ):
+                explainer = shap.explainers.Additive(
+                    raw_model, maskers, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(sampled_X)
+            else:
+                # fallback to default explainer
+                explainer = shap.Explainer(
+                    predict_fn, sampled_X, feature_names=truncated_feature_names
+                )
+                shap_values = explainer(sampled_X)
+
+        _logger.info(f'Shap explainer {explainer.__class__.__name__} is used.')
+
+        # TODO: seems infer pip req fail when log_explainer.
+        # TODO: The explainer saver is buggy, if `get_underlying_model_flavor` return "unknown",
+        #   then fallback to shap explainer saver, and shap explainer will call `model.save`
+        #   for sklearn model, there is no `.save` method, so error will happen.
```

Comment:
Makes sense, probably hard to fix the bug in SHAP quickly. Can we try / catch this?

Reply:
There's some discussion about the issue here:
https://github.com/mlflow/mlflow/pull/5092#discussion_r770209440

* Event if log explainer successfully but it may still fail when loading back (due to shap's buggy loader)